Create replication.md quick walkthrough for pg_tde

Andriciuc opened the pull request:
started to populate replication with information and added to ToC

Andriciuc committed:
updated replication.md

started to populate replication with information and added to ToC

Andriciuc committed:
Update replication.md

Andriciuc committed:
small fixes to text

Updated text a bit, need to further lint it more

PG-802 linked

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Updated replication.md

Replaced master with principal

Andriciuc committed:
Merge branch 'DOCS-create-replication.md' of https://github.com/percona/postgres into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
-<!-- To add information on Replication affecting pg_tde for customers and how it interacts with it
-Also, don't forget to add it to TOC>
\ No newline at end of file
+# Streaming Replication with pg_tde (heap_tde)
```
tde_heap

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
\ No newline at end of file
+# Streaming Replication with pg_tde (heap_tde)
+
+This section outlines how to set up PostgreSQL streaming replication when Percona's `pg_tde` extension (specifically the `heap_tde` access method) is enabled on the primary server.
```
tde_heap

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+This section outlines how to set up PostgreSQL streaming replication when Percona's `pg_tde` extension (specifically the `heap_tde` access method) is enabled on the primary server.
+
+!!! note
+    Physical streaming replication copies data blocks directly from the primary to the standby. If these blocks are encrypted on the primary, they arrive **encrypted** on the standby. Therefore, the standby **must** have access to the **exact same encryption keys** as the primary to decrypt and read this data.
```
Generally, this is not true for multiple reasons:

* We have two level of keys (principal, internal). Internal keys get copied during replica creation, but to use them, the replica needs access to the same principal keys - so that part is correct
* Later when users change the principal keys / keyring configuration, it gets changed on all servers - so while it is true that the replica needs access to the same principal key, it's handled automatically
* Replication uses WAL, and we write the unencrypted page data into WAL. (unless we have wal encryption setup, but in that case, WAL is encrypted with its own key). So the encryption of the data blocks doesn't interact replication
* Internal keys can be different for tables created after the replica is setup, or tables that were altered / truncated / ... after the replica was setup. Initially they will be the same on the primary and replica because replica creation physically copy the files, but later different servers can create different keys for new database files.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+## Concepts
+
+* **`pg_tde`:** Percona's extension for Transparent Data Encryption in PostgreSQL.
+* **`heap_tde`:** The table access method provided by `pg_tde` that encrypts entire tables using a principal key.
```
tde_heap

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+* **`pg_tde`:** Percona's extension for Transparent Data Encryption in PostgreSQL.
+* **`heap_tde`:** The table access method provided by `pg_tde` that encrypts entire tables using a principal key.
+* **principal Key:** The central key used by `pg_tde` to encrypt/decrypt data. Identified by a `pg_tde_principal_key_name`.
+* **Key Provider:** The mechanism `pg_tde` uses to fetch the principal key (e.g., local files, HashiCorp Vault). Configured via `pg_tde_key_provider`.
```
There's no method named `pg_tde_key_provider`, only multiple methods with somewhat similar names.

We also support KMIP

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+The standby server(s) **must** be configured to use the **same key provider settings** and access the **same principal key (by name and content)** as the primary server.
+
+### Scenario 1: File-based Key Provider
```
This won't work:

in practice, the primary and the replica are on different machines, so even if the filenames are the same, they refer to different files.

And even if they use a shared filesystem to access the same files, pg_tde assumes that the specified file is exclusive to the current server, so it won't work.

In practice:
* `keyring_file` is only for development and testing purposes, we shouldn't even try to document advanced use cases for it
* `keyring_file` doesn't work with replication other than the most simple case, where users **never** change principal keys

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+### Scenario 2: HashiCorp Vault Key Provider**
+
+1. **Primary:**
+    * Ensure the principal key (e.g., `my_vault_principal_key`) exists in Vault at the configured path (for example, `secret/data/postgres/keys`).
```
No, `pg_tde` will create the key

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+1. **Primary:**
+    * Ensure the principal key (e.g., `my_vault_principal_key`) exists in Vault at the configured path (for example, `secret/data/postgres/keys`).
+    * In `postgresql.conf`:
```
These settings do not exists at all, the only thing required in the config is `shared_preload_libraries`

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+        pg_tde_vault_token = 'your-vault-access-token' # Or use other auth methods
+        pg_tde_vault_mount_path = 'secret/' # Path where KV engine is mounted
+        # pg_tde_vault_ca_path = '/path/to/vault_ca.crt' # If using self-signed certs
+        pg_tde_principal_key_name = 'my_vault_principal_key' # Key name within Vault path
```
Where did you get these settings from? Some of these never existed, even in earlier versions, so it can't be from earlier documentation...

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,257 @@
+        ```sql
+        CREATE EXTENSION IF NOT EXISTS pg_tde;
+        -- Add the key definition (pg_tde will fetch it from Vault)
+        SELECT pg_tde_principal_key_add('my_vault_principal_key', NULL, 'vault');
````
parameters here completely wrong, please see other sections which use this function

also call to provider creation missing before this.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+        CREATE EXTENSION IF NOT EXISTS pg_tde;
+        -- Add the key definition (pg_tde will fetch it from Vault)
+        SELECT pg_tde_principal_key_add('my_vault_principal_key', NULL, 'vault');
+        -- Set this key as the active principal key
+        SELECT pg_tde_set_principal_key('my_vault_principal_key', 'vault');
```
names of the set functions were changed, this has to be local/global/default/.... please check the relevant documentation

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,257 @@
+        ```
+
+2. **Standby:**
+    * In the standby's `postgresql.conf`, configure `pg_tde_key_provider` and all `pg_tde_vault_*` parameters, and `pg_tde_principal_key_name` to be **identical** to the primary's configuration.
````
again, there's no such configuration

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+2. **Standby:**
+    * In the standby's `postgresql.conf`, configure `pg_tde_key_provider` and all `pg_tde_vault_*` parameters, and `pg_tde_principal_key_name` to be **identical** to the primary's configuration.
+    * Ensure the standby server has network access to the Vault server and the necessary credentials/permissions to read the *same principal key* from Vault.
```
network access is required, but the credentials are configured in pg_tde, so they will be the same as the primary, automatically...

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+**Is it possible to use the same key provider and principal key?**
+
+**Yes, it is mandatory.** For physical replication to function with TDE, the standby *must* use the same principal key (by name and content, accessed via the same provider configuration) as the primary to decrypt the replicated data blocks.
```
streaming replication. Has nothing to do with replicated data blocks.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+## Encrypted Table Replication
+
+**If a table on the primary site is encrypted, will all the standby sites also have the encrypted table replicated, given that the same key is implemented on the standby sites as well?**
```
> , given that the same key is implemented on the standby sites as wel

this part doesn't make any sense

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+Yes.
+
+Physical replication works at the block level.
```
Should this be a heading? Also, the description in this section is completely incorrect.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+If the standby lacks the correct key, replicated data remains encrypted and unreadable, and queries will fail.
+
+## Setup Steps
```
We had setup instructions with two different providers above, why is another setup section here now?

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,257 @@
+2. **Configure `postgresql.conf` for `pg_tde`:**
+
+    ```ini
+    shared_preload_libraries = 'pg_tde' // Add pg_tde, comma-separated if others exist
````
The ini file uses `#` for comments

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,257 @@
+    ```ini
+    shared_preload_libraries = 'pg_tde' // Add pg_tde, comma-separated if others exist
+
+    # Example for File Provider:
````
Again completely fantasy, these settings do not exists.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+    wal_level = replica
+    max_wal_senders = 5
+    archive_mode = on # Recommended
+    archive_command = 'cp %p /path/to/archive/%f' # Example
```
shared_preload_libraries missing. And these settings have nothing to do with pg_tde, I don't think we should mention them.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+4. **Prepare principal Key in Provider:**
+
+    * **File:** Create the directory specified in `pg_tde_file_keys_path`. Create the key file inside this directory, named as per `pg_tde_principal_key_name`.
```
Again not how this works.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,257 @@
+        sudo -u postgres chmod 600 /var/lib/postgresql/pg_tde_keys/my_principal_key
+        ```
+
+    * **Vault:** Ensure the key exists in Vault at the correct path and is accessible.
````
Again not how it works, pg_tde creates the keys needed.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+
+### B. Standby Server Setup**
+
+1. **Install `pg_tde` extension files.**
```
The first step should be "create the replica using basebackup or something", please check/try how replication setup works with postgres.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,257 @@
+        (2 rows)
+        ```
+
+    * **Negative Test (Optional):** Temporarily make the principal key file inaccessible on the standby (e.g., rename `pg_tde_file_keys_path` directory or the key file within it) or revoke Vault token. Restart standby. Queries to `encrypted_data` should now fail with a TDE key access error. Restore access and restart to confirm functionality.
````
Not possible with the way pg_tde actually works.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,257 @@
+* **principal Key Security:** The principal key is paramount. Securely back it up. For file provider, protect the key file and its directory stringently.
+* **Identical Configuration:** All `pg_tde_*` GUCs relevant to key provider and key name **must** be identical on primary and all standbys.
+* **Key Rotation (`pg_tde_rotate_principal_key`):**
+    1. Create/place the new key in your chosen key provider (new file in `pg_tde_file_keys_path` or new key in Vault).
```
Again this is not how key rotation works.

Andriciuc committed:
updated replication.md with proper information

Initial commit was messy, reworded for basic users and cleared up any confusing parts like heap_tde which was incorrect

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,152 @@
+    max_replication_slots = 1
+    wal_keep_size        = '1GB'
+    # Enable TDE in WAL pipeline
+    shared_preload_libraries = 'pg_tde' # Loads TDE hooks at server start.
```
This is the only setting that's required, everything above is optional. Those either repeat the default value, or make things even more restricted than by default.

Andriciuc replied:
Thank you! Will update to remove optionals. Also, please see below, possibly keeping them (since it doesn't "hurt"?)

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+    ```
+
+    !!! note
+        Make sure you set `max_replication_slots` before creating any slot.
````
Why? By default it is 10, in this example you set it to 1.

Andriciuc replied:
> Why? By default it is 10, in this example you set it to 1.

For simplicity I wanted the examples to reflect what a new user goes through, if starting from 1 replica (having the possibility to increase it whenever they want depending on topology). For sure we can remove the other replication GUCs, but... as a new user I would appreciate having a full example. Thoughts?

dutow replied:
You can set up 1 replica even when this setting is its default value (10). It's not the required number, it's the maximum number supported by the current configuration. There's no need to change this setting, unless you want more than 10 replicas.

Andriciuc replied:
Removed the note and updated max_replication_slots = 10 # the default value

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,152 @@
+    !!! note
+        Make sure you set `max_replication_slots` before creating any slot.
+
+* (Optional - **To review**) Create a physical slot to retain the encrypted WAL, set `max_replication_slots ≥ 1` and then:
```
Why? basebackup can handle the slots with `-C`.

Andriciuc replied:
Removed the following as -c fast was added:

* (Optional - **To review**) Create a physical slot to retain the encrypted WAL, set `max_replication_slots ≥ 1` and then:

    ```sql
    SELECT pg_create_physical_replication_slot('tde_slot');
    ```

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+    SELECT pg_create_physical_replication_slot('tde_slot');
+    ```
+
+* Ensure you have [configured the global key provider](global-key-provider-configuration/index.md).
````
Why global? Database providers work too...

Andriciuc replied:
Updated to: Ensure you have configured the provider.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,152 @@
+
+* Ensure you have [configured the global key provider](global-key-provider-configuration/index.md).
+* Create the [principal key](functions#pg_tde_set_server_key_using_global_key_provider).
+* Enable [WAL encryption](wal-encryption). **Restart** PostgreSQL.
```
WAL encryption is beta and still has some issues with replication, we shouldn't recommend it in our documentation.

Andriciuc replied:
removed the WAL statement

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,152 @@
+* Ensure you have [configured the global key provider](global-key-provider-configuration/index.md).
+* Create the [principal key](functions#pg_tde_set_server_key_using_global_key_provider).
+* Enable [WAL encryption](wal-encryption). **Restart** PostgreSQL.
+* Ensure the extension is installed in each database:
```
Where it's needed.

Andriciuc replied:
Updated to: 

* Ensure the extension is installed where it is needed:

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+    CREATE EXTENSION IF NOT EXISTS pg_tde;
+    ```
+
+* (Optional) If you want to block any unencrypted tables, you can [enforce table-level encryption](variables#pg_tde.enforce_encryption):
````
This and some other information here - again I don't see how is this related to replication?

Andriciuc replied:
Removed

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+```
+
+!!! note
+    Run pg_basebackup only **after** slot creation if using `--slot=tde_slot`.
````
Why?

Andriciuc replied:
Removed, not needed

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+```
+
+!!! note
+    By default `pg_tde.inherit_global_providers = on`, so the standby inherits your global KMS configuration automatically.
````
how is this note relevant to the current documentation section?

Andriciuc replied:
it is not, removed!

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+* Install the extension so the standby can register `tde_heap`:
+
+```sql
+CREATE EXTENSION IF NOT EXISTS pg_tde;
````
where and how? you don't have to do this on the replica...

Andriciuc replied:
Removed as it is not necessary:

* Install the extension so the standby can register `tde_heap`:

```sql
CREATE EXTENSION IF NOT EXISTS pg_tde;
```

!!! note
    Ensure the standby has access to the **same** encryption key material or provider configuration used by the primary.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+CREATE EXTENSION IF NOT EXISTS pg_tde;
+```
+
+* For PostgreSQL ≥13, in `postgresql.auto.conf`, ensure you have set the following:
````
you don't have to set these, again, just use basebackup

Andriciuc replied:
removed this

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,152 @@
+pg_basebackup \
+-h primary_ip \
+-D /var/lib/pgsql/data \
+-U example_replicator \
```
`-C`

Andriciuc replied:
added -c fast for checkpoints:

export PGPASSWORD='example_password'
pg_basebackup \
  -h primary_ip \
  -D /var/lib/pgsql/data \
  -U example_replicator \
  --wal-method=stream \
  --slot=tde_slot \
  -c fast \
  -v -P

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,152 @@
+### On the standby host, after pg_basebackup and configuration is set:
+
+```bash
+touch $PGDATA/standby.signal
````
this again shouldn't be necessary

Andriciuc replied:
removed the following:

### On the standby host, after pg_basebackup and configuration is set:

```bash
touch $PGDATA/standby.signal
```

PostgreSQL looks for the `standby.signal` (replacing `recovery.conf` as of v12) to know it should enter streaming recovery.

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Update replication.md

updated with added -c fast option for pg_basebackup and implemented Zsolt's feedback

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,104 @@
+
+The following steps assume:
+
+* You have configured a global key provider for both **primary** and **standby** (see [Configure Key Management (KMS)](global-key-provider-configuration/index.md)).
```
you don't have to configure it for the standby, it's automtatic (other than that the certificate files has to be accessible for the standby too, and you have to add pg_tde to the shared preload libraries)

Andriciuc replied:
Removed the standby assumption and added the following line: 

Ensure the certificate files are accessible for the standby, and that `pg_tde` is added to the shared preload libraries.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,104 @@
+The following steps assume:
+
+* You have configured a global key provider for both **primary** and **standby** (see [Configure Key Management (KMS)](global-key-provider-configuration/index.md)).
+* You have enabled `pg_tde` and you have setup at least one active key on the **primary**.
```
Again wrong order, you have to enable pg tde first.

And then the previous sentence was about key setup...

Andriciuc replied:
moved the order, thank you!

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,104 @@
+
+* You have configured a global key provider for both **primary** and **standby** (see [Configure Key Management (KMS)](global-key-provider-configuration/index.md)).
+* You have enabled `pg_tde` and you have setup at least one active key on the **primary**.
+* Both primary and standby run the **same** Percona PostgreSQL version.
```
This isn't tde specific. And minor version differences should work, so not even 100% true

Andriciuc replied:
Great! Removed this bit

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,104 @@
+    ```ini
+    # Example of WAL and replication settings
+    wal_level            = replica
+    max_wal_senders      = 5
````
Again, please do not add changes to random unrelated parameters, such as the number of wal senders, it has nothing to do with tde.

Andriciuc replied:
Hmm, actually, perhaps this is unnecessary. I decided to remove the conf and left just this:

* Ensure you have configured `postgresql.conf`.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,104 @@
+  -D /var/lib/pgsql/data \
+  -U example_replicator \
+  --wal-method=stream \
+  --slot=tde_slot \
```
The `-C` parameter for creating a slot is missing

Andriciuc replied:
added -C \ after slot=tde_slot

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Update replication.md

updated based on feedback

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,94 @@
+
+* Ensure you have configured `postgresql.conf`.
+* Ensure you have configured the provider.
+* Create the [principal key](functions#pg_tde_set_server_key_using_global_key_provider).
```
In the assumptions above, `and you have setup at least one active key on the ` - if we assume it was done already, why do we have to do it again?

Andriciuc replied:
Fixed! User should already have an active key enabling tde, removed that line

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,94 @@
+The following steps assume:
+
+* You have enabled `pg_tde` and you have setup at least one active key on the **primary**.
+* You have configured a global key provider for the **primary**, see [Configure Key Management (KMS)](global-key-provider-configuration/index.md) for more information.
```
Can be any key provider, it is not required to be global.

Andriciuc replied:
removed the global part: ....configure a key provider on the primary...

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,94 @@
+### Configure postgresql.conf
+
+* Ensure you have configured `postgresql.conf`.
+* Ensure you have configured the provider.
```
In the assumptions above "You have configured a global key provider"  - if we assume it's already done, why do we configure it again?

Andriciuc replied:
True! Removed

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,94 @@
+* Ensure the extension is installed where it is needed:
+
+    ```sql
+    CREATE EXTENSION IF NOT EXISTS pg_tde;
````
We can't create keys or providers without installing the extension, this should be the second step

Andriciuc replied:
moved this as the second step to create pg_tde and then create the key for it

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,94 @@
+
+### Configure pg_hba.conf
+
+To connect to the replication server, add the following line in `pg_hba.conf`:
```
Maybe to allow the replica to connect to the primary server, or something like that?

Andriciuc replied:
reworded for better clarity: To allow the replica to connect to the primary server, add the following line in `pg_hba.conf`:...

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,94 @@
+
+## 2. Configure the Standby
+
+### Perform an encrypted database backup
```
I would call it simply database backup

Andriciuc replied:
Reworded to: 

### Perform a database backup

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,94 @@
+
+```ini
+shared_preload_libraries = 'pg_tde'
+hot_standby = on
````
this is the default value for hot standby.

Also, we do not mention `shared_preload_libraries` for the primary, we should either mention it for both or none.

Andriciuc replied:
I added shared preload lib to both primary and secondary with a variation of:

 Ensure that the following line is present in shared_preload_libraries...

```ini
shared_preload_libraries = 'pg_tde'
```

Andriciuc committed:
Update replication.md

updated with feedback, removed global for configuring key providers, reordered the setup to ensure pg_tde is enabled and THEN create a key, small clarification updates

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -4,23 +4,28 @@ This section outlines how to set up PostgreSQL streaming replication when the `p
-* You have enabled `pg_tde` and you have setup at least one active key on the **primary**.
-* You have configured a global key provider for the **primary**, see [Configure Key Management (KMS)](global-key-provider-configuration/index.md) for more information.
+* You have enabled `pg_tde`.
+* You have configured a key provider for the **primary**, see [Configure Key Management (KMS)](global-key-provider-configuration/index.md) for more information.
```
This assumption still conflicts with later steps - to create a key provider, you have to (1) add pg_tde to the shared preload libraries, and (2) create the extension

Both are later steps.

Same with the "you have enabled pg_tde" - the above 2 steps are the same.

Andriciuc replied:
Hmm, I removed these "considerations"and replaced with the following paragraph:

The following steps assume you have [installed](install.md) and [configured](setup.md) `pg_tde`.

Perhaps this is also not required, but it is worth keeping just to emphasize the feature works only with tde.

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Update replication.md

removed considerations and added a paragraph directing the user to install and config, also added a link for tde_heap docs

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,94 @@
\ No newline at end of file
+This section outlines how to set up PostgreSQL streaming replication when the `pg_tde` extension (specifically the [`tde_heap`](index/table-access-method.md) access method) is enabled on the primary server.
+
+The following steps assume you have [installed](install.md) and [configured](setup.md) `pg_tde`.
```
Still the same conflict as before - `setup.md` tells the user to add it to `shared_preload_libraries` and create the extension. On this page we tell them to do it again.

Also unclear of where - on the primary? On the replica? On both? Because you have to add it to the `shared_preload_libraries` on both, but you only have to `CREATE EXTENSION` on the primary.

Andriciuc replied:
Very good feedback here, in fact I have reworded this whole intro section. I have done the following:

- Reworded the second paragraph and removed the install.md reference. I also added (which includes...) to ensure that experienced users don't just ignore the setup link and have a proper quick view of the prerequisites as well as helping new users to reinforce what they need to have configured already.
- Removed CREATE EXTENSION from step 1 Configure the Primary, it is redundant. I also updated the principal key function step with a fixed link
- Updated primary replication role paragraph
- Removed initial standby setup as we already cover this
- Updated step 3 Start and Validate replication with an intro text and added a tip to remind the user they can verify if the encrypt is

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Update replication.md

Updated the following:
- Reworded the second paragraph and removed the install.md reference. I also added (which includes...) to ensure that experienced users don't just ignore the setup link and have a proper quick view of the prerequisites as well as helping new users to reinforce what they need to have configured already.
- Removed CREATE EXTENSION from step 1 Configure the Primary, it is redundant. I also updated the principal key function step with a fixed link
- Updated primary replication role paragraph
- Removed initial standby setup as we already cover this
- Updated step 3 Start and Validate replication with an intro text and added a tip to remind the user they can verify if the encrypt is

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'release-17.5.2' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'release-17.5.2' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'release-17.5.2' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'release-17.5.2' into DOCS-create-replication.md

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,86 @@
+
+Before you begin, ensure you have followed the [`pg_tde` setup instructions](setup.md), which includes:
+
+- Installing the `pg_tde` extension binaries, where they are needed, on **both** the primary and standby servers.
```
both -> all, as replication can use any number of servers.

Andriciuc replied:
Replaced both with all: ... , on **all** servers.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,86 @@
+Before you begin, ensure you have followed the [`pg_tde` setup instructions](setup.md), which includes:
+
+- Installing the `pg_tde` extension binaries, where they are needed, on **both** the primary and standby servers.
+- Configuring `shared_preload_libraries = 'pg_tde'` in `postgresql.conf` on **both** systems.
```
both -> all, as replication can use any number of servers.

Andriciuc replied:
Replaced both with all: ... , on **all** systems.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
```diff
@@ -1,4 +1,86 @@
+
+## 1. Configure the Primary
+
+### Create a principal key
```
The assumptions above still include 

> setting a principal key on the **primary**.

Andriciuc replied:
Removed "...and setting a principal key on the **primary**."

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,86 @@
+SELECT pg_reload_conf();
+```
+
+## 2. Configure the Standby
````
Again an order issue: the assumptions above says that you already added pg_tde to the shared_preload_libraries on the standby: 

> - Configuring `shared_preload_libraries = 'pg_tde'` in `postgresql.conf` on **both** systems.

But you can't do that before this step, as you create the data directory in this step.

Andriciuc replied:
I have removed the  "which includes" lines, superfluous as we mention pg tde setup already, and added an additional step for the standby after backup is done:

### Configure postgresql.conf

After the base backup completes, add the following line to the standby's `postgresql.conf` file:

```ini
shared_preload_libraries = 'pg_tde'
```

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,86 @@
+Start the PostgreSQL service:
+
+```bash
+sudo systemctl start postgresql
````
This (use of the default systemctl) assumes that you run the primary and replica on different servers - which makes sense in production situations.

However, in this case an important step/information is missing:

* you can't use keyring_file, as it stores information in a local file, and doesn't expect concurrent access to that file
* for vault/kmip, both the primary and the replica needs access to the same configuration files, at the same location. For example if you configure vault with the secret file `/path/to/the/secret.file' (and optionally some CA file), that file has to exist exactly at the specified location on both the primary and the replica. Same with files used by the kmip configuration.

Andriciuc replied:
To integrate this idea, I rewrote the intro of the step, I added the assumption that the user uses primary/standy separately and added a warning regarding kmip configs:

## 3. Start and validate replication

Assuming that the primary and the standby are running on separate hosts, start the PostgreSQL service:

```bash
sudo systemctl start postgresql
```

!!! warning "Key management consistency **required** for replication"

    If you're using a KMS provider, such as Vault or KMIP, make sure that both the primary and the standby have access to the **same** key management configuration, and that the paths to the configuration files are identical on both systems.

    For example:
    - If you configure Vault with a secret path: `/path/to/secret.file`, then that file **must** exist at the same path on both the primary and the standby.
    - If you use the `keyring_file` provider, be aware that it stores key material in a local file and it is **not designed** for shared or concurrent use across multiple servers. It is **not recommended** in replication setups.

dutow commented on contrib/pg_tde/documentation/docs/replication.md:
````diff
@@ -1,4 +1,86 @@
+```
+
+!!! tip
+    Want to verify if everything works? Run `SELECT pg_tde_is_encrypted('your_encrypted_table');` on the standby to confirm that the encryption is active and the keys are resolved.
````
This needs the creation of an encrypted table first.

Andriciuc replied:
Updated tip with the following:

...works? After creating an encrypted table on the primary, run the following command on the standby to confirm that the encryption is active and the keys are resolved:....

Andriciuc committed:
updated based on latest feedback

- updated intro to remove () and replace with , , for tde heap access method mention
- removed assumption steps when talking about pg_tde setup
- added Configure postgresql.conf step for configuring the standby after db backup
- updated step 3 with clear assumption that primary and standby run separately and added a warning regarding Key Management consistency between replica and primary
- updated tip to include a way to copy the command with one click and added that after creating an encrypted table you can use it

Andriciuc committed:
Merge branch 'release-17.5.2' into DOCS-create-replication.md

Andriciuc committed:
Merge branch 'release-17.5.2' into DOCS-create-replication.md
